server/src/main/java/org/opensearch/tasks/TaskResourceTrackingService.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.tasks;
+
+import com.sun.management.ThreadMXBean;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ConcurrentCollections;
+import org.opensearch.common.util.concurrent.ConcurrentMapLong;
+import org.opensearch.threadpool.RunnableTaskExecutionListener;
+
+import java.lang.management.ManagementFactory;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.opensearch.tasks.ResourceStatsType.WORKER_STATS;
+
+/**
+ * Service that helps track resource usage of tasks running on a node.
+ */
+public class TaskResourceTrackingService implements RunnableTaskExecutionListener {
+
+    public static final Setting<Boolean> TASK_RESOURCE_TRACKING_ENABLED = Setting.boolSetting(
+        "task_resource_tracking.enabled",
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private static final ThreadMXBean threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
```

Comment:
Should we abstract this out to a separately, such that ``TaskResourceTrackingService`` is not necessarily aware of the underneath implementation for ResourceUsageMetrics?

Reply:
Thought of doing this when logic was in TaskManager but given Resource tracking service is carved out to handle all related things skipped it for now. I see this being used in similar way across the codebase. What we can do is to have different refactor PR if required in which we abstracts out which exact implementation we are using to fetch resource metrics. Thoughts?

Reply:
Yeah, I would love to see that. Definitely not a blocker for this PR. We can do that as a follow-up